components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/client/authentication/OAuthClientAuthnService.java

```diff
@@ -258,4 +295,245 @@ private boolean canAuthenticate(OAuthClientAuthenticator oAuthClientAuthenticato
+                    OAuth2ErrorCodes.INVALID_REQUEST, e);
+        }
+    }
+
+    /**
+     * Obtain the client authentication method registered for the application.
+     *
+     * @param clientId     Client ID of the application.
+     * @return Registered client authentication method for the application.
+     * @throws OAuthClientAuthnException
+     */
+    public String retrieveRegisteredAuthMethod(String clientId) throws OAuthClientAuthnException {
+
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (StringUtils.isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (OAuthConstants.TOKEN_ENDPOINT_AUTH_METHOD.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return OAuthConstants.NOT_APPLICABLE;
+    }
+
+    /**
+     * Check whether the application should support FAPI.
+     *
+     * @param clientId       Client ID of the application.
+     * @return Whether the application should support FAPI.
+     * @throws OAuthClientAuthnException
+     */
+    private String isFapiApplication(String clientId) throws OAuthClientAuthnException {
+
+        String isFapiApp = "IsFAPIApp";
+        try {
+            ServiceProvider serviceProvider = OAuth2Util.getServiceProvider(clientId);
+            if (StringUtils.isEmpty(serviceProvider.getCertificateContent())) {
+                ServiceProviderProperty[] serviceProviderProperties = serviceProvider.getSpProperties();
+                for (ServiceProviderProperty serviceProviderProperty : serviceProviderProperties) {
+                    if (isFapiApp.equals(serviceProviderProperty.getName())) {
+                        return serviceProviderProperty.getValue();
+                    }
+                }
+            }
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException("Token signing algorithm not registered",
+                    OAuth2ErrorCodes.INVALID_REQUEST);
+        }
+        return null;
+    }
+
+    /**
+     * Obtain the client ID of the application from the request.
+     *
+     * @param request       Http servlet request.
+     * @return Client ID of the application.
+     * @throws OAuthClientAuthnException
+     */
+    private String extractClientId(HttpServletRequest request, Map<String, List> contentParams)
```

Comment:
This method is not readable and too long. Shall we break it into small methods? We can extract into some simple methods as `getClientIdFromRequestParams(request)` , `getClientIdFromBody(contentParams)`, ` getClientIdFromBasicAuth(request)`

Reply:
Fixed in 36cde69f01a0865dc525ff8c963f66e0af53cf89